### PR TITLE
Move StateAllocationFilter to Stable

### LINKED ISSUE #3096: Graduate `StateAllocationFilter` to Stable
I propose that we graduate the `StateAllocationFilter` feature to stable. It has been in beta since 1.26.0. 

Context:
https://development.agones.dev/site/docs/guides/feature-stages/#feature-gates

### PR DESCRIPTION
**Pull request by Kalaiselvi84**
**What type of PR is this?**
> Uncomment only one ` /kind <>` line, press enter to put that in a new line, and remove leading whitespace from that line:
>
> /kind breaking
> /kind bug
> /kind cleanup
> /kind documentation

/kind feature

> /kind hotfix

**What this PR does / Why we need it**:

**Which issue(s) this PR fixes**:

Closes #3096

**Special notes for your reviewer**:

### CODE REVIEW

**Comment by zmerlynn** [comment]
Moving to `Stable` is more involved. You basically need to go through e.g. https://github.com/search?q=repo%3Agoogleforgames%2Fagones%20StateAllocationFilter&type=code and make sure:
* anything conditional on the feature gate is folded to true - this often comes with opportunities for refactoring
* and the gate should no longer be called out in the docs, as it's fully incorporated

File: pkg/allocation/converters/converter_test.go [hunk 1]
```diff
@@ -144,7 +144,7 @@ func TestConvertAllocationRequestToGameServerAllocation(t *testing.T) {
 		},
 		{
 			name:     "all fields are set",
-			features: fmt.Sprintf("%s=false&%s=false", runtime.FeaturePlayerAllocationFilter, runtime.FeatureStateAllocationFilter),
+			features: fmt.Sprintf("%s=false", runtime.FeaturePlayerAllocationFilter),
```
Comment (zmerlynn) [inline pkg/allocation/converters/converter_test.go:147]: For the `false` cases, you should just remove the tests.

File: pkg/gameserverallocations/find_test.go [hunk 2]
```diff
@@ -127,7 +127,7 @@ func TestFindGameServerForAllocationPacked(t *testing.T) {
 				assert.Equal(t, ErrNoGameServer, err)
 				assert.Nil(t, gs)
 			},
-			features: fmt.Sprintf("%s=false&%s=false", runtime.FeaturePlayerAllocationFilter, runtime.FeatureStateAllocationFilter),
+			features: fmt.Sprintf("%s=false", runtime.FeaturePlayerAllocationFilter),
```
Comment (zmerlynn) [inline pkg/gameserverallocations/find_test.go:130]: Ditto here - remove the "false" cases (you can't disable it anymore)

File: site/content/en/docs/Integration Patterns/high-density-gameservers.md [hunk 3]
```diff
@@ -6,7 +6,7 @@ weight: 70
 description: >
   How to run multiple concurrent game sessions in a single GameServer process.
 ---
-{{< beta title="Allocation State Filter" gate="StateAllocationFilter" >}}
+{{< stable title="Allocation State Filter" gate="StateAllocationFilter" >}}
```
Comment (zmerlynn) [inline site/content/en/docs/Integration Patterns/high-density-gameservers.md:9]: Here we can just take the line away - they can't disable the gate anyways.

Reminder, though, that docs changes need a release guard: https://agones.dev/site/docs/contribute/#within-a-page

**Review by zmerlynn — COMMENTED** [review]
Much closer.

File: pkg/allocation/converters/converter_test.go (hunk 1 above)
Comment (Kalaiselvi84) [reply pkg/allocation/converters/converter_test.go:147]: sorry I am not getting it, do I need to remove this block https://github.com/googleforgames/agones/blob/main/pkg/allocation/converters/converter_test.go#L217-L317 from the function `TestConvertAllocationRequestToGameServerAllocation`?

File: pkg/gameserverallocations/find_test.go (hunk 2 above)
Comment (Kalaiselvi84) [reply pkg/gameserverallocations/find_test.go:130]: The expectation is that these lines will be removed https://github.com/googleforgames/agones/blob/main/pkg/gameserverallocations/find_test.go#LL96C6-L96C6 ?

File: pkg/allocation/converters/converter_test.go (hunk 1 above)
Comment (zmerlynn) [reply pkg/allocation/converters/converter_test.go:147]: Yes, and others cases below this. The `features` string is setting the features flag, but on Stable promotion the "=false" setting no longer exists (the feature no longer exists, it's fully incorporated into the product).

**Comment by Kalaiselvi84** [comment]
Are there any changes that need to be made to the following files,

https://github.com/googleforgames/agones/blob/main/site/content/en/docs/Reference/gameserverallocation.md#this-is-useful-for-things-like-smoke-testing-of-new-game-servers

https://github.com/googleforgames/agones/blob/main/site/content/en/docs/Reference/agones_crd_api_reference.html#L532-L539

**Comment by zmerlynn** [comment]
@Kalaiselvi84 Yes, the first one should remove references to the feature (no need to document a feature gate that doesn't exist). The second one:

> https://github.com/googleforgames/agones/blob/main/site/content/en/docs/Reference/agones_crd_api_reference.html#L532-L539

is auto-generated by `make gen-api-docs`, which you'll need to run.

Also @agones-bot is complaining about `lint` - I highly recommend running `make lint` before iterating on the PR, and probably `make test-go` at the very least.

**Comment by Kalaiselvi84** [comment]
> @Kalaiselvi84 Yes, the first one should remove references to the feature (no need to document a feature gate that doesn't exist). The second one:
> 
> > https://github.com/googleforgames/agones/blob/main/site/content/en/docs/Reference/agones_crd_api_reference.html#L532-L539
> 
> is auto-generated by `make gen-api-docs`, which you'll need to run.
>
> Also @agones-bot is complaining about `lint` - I highly recommend running `make lint` before iterating on the PR, and probably `make test-go` at the very least.

`make test-go` is not succeeded. I have attached the log info in the buganizer ticket[b/282777251]

File: pkg/allocation/converters/converter_test.go [hunk 4]
```diff
@@ -215,132 +215,8 @@ func TestConvertAllocationRequestToGameServerAllocation(t *testing.T) {
 			},
 		},
 		{
-			name:     "all fields are set",
```
Comment (zmerlynn) [inline pkg/allocation/converters/converter_test.go:218]: I'm sorry, I answered too quickly on this test. I think the nature of this test was to set all the built-in fields, but I think it was disabling all of the features to minimize the changes. Both this and the previous probably need to be restored.

Unfortunately, this part is going to require a bit deeper understanding of the test. You'll probably have to change this to just PlayerAllocationFilter=false and CountsAndList=false, then iterate on `want` until the expected output is correct. You can generally assume that the current code is correct and adjust the output to match.

File: site/content/en/docs/Guides/feature-stages.md [hunk 5]
```diff
@@ -47,11 +46,10 @@ The current set of `alpha` and `beta` feature gates:
...
-| [Reset Metric Export on Fleet / Autoscaler deletion]({{% relref "./metrics.md#dropping-metric-labels" %}})            | `ResetMetricsOnDelete`         | Enabled  | `Beta`  | 1.26.0 |
 | [Graceful Termination for GameServer SDK](https://github.com/googleforgames/agones/pull/2205)                         | `SDKGracefulTermination`       | Enabled  | `Beta`  | 1.18.0 |
-| [GameServer state filtering on GameServerAllocations](https://github.com/googleforgames/agones/issues/1239)           | `StateAllocationFilter`        | Enabled  | `Beta`  | 1.26.0 |
 | [GameServer player capacity filtering on GameServerAllocations](https://github.com/googleforgames/agones/issues/1239) | `PlayerAllocationFilter`       | Disabled | `Alpha` | 1.14.0 |
 | [Player Tracking]({{< ref "/docs/Guides/player-tracking.md" >}})                                                      | `PlayerTracking`               | Disabled | `Alpha` | 1.6.0  |
+| [Reset Metric Export on Fleet / Autoscaler deletion]({{% relref "./metrics.md#dropping-metric-labels" %}})            | `ResetMetricsOnDelete`         | Disabled | `Alpha` | 1.26.0 |
```
Comment (zmerlynn) [inline site/content/en/docs/Guides/feature-stages.md:52]: I think you accepted the wrong side of a merge conflict here - `ResetMetricsOnDelete` should be in `Beta` as of 1.32 yeah?

File: pkg/gameserverallocations/find_test.go [hunk 6]
```diff
@@ -82,53 +82,6 @@ func TestFindGameServerForAllocationPacked(t *testing.T) {
...
-				assert.Equal(t, gs, list[index])
-				assert.Equal(t, agonesv1.GameServerStateReady, gs.Status.State)
-
-				list = nil
-				gs, _, err = findGameServerForAllocation(gsa, list)
-				assert.Error(t, err)
-				assert.Equal(t, ErrNoGameServer, err)
-				assert.Nil(t, gs)
-			},
-			features: fmt.Sprintf("%s=false&%s=false", runtime.FeaturePlayerAllocationFilter, runtime.FeatureStateAllocationFilter),
```
Comment (zmerlynn) [inline pkg/gameserverallocations/find_test.go:130]: yeah this is another one I was probably too quick on, sorry. I think `StateAllocationFilter` needs to be taken out, but you'll have to readjust the test.

**Review by zmerlynn — COMMENTED** [review]
For the tests that are failing, you'll probably have to change the expected values to match the values we're seeing. I haven't run through them carefully yet, though, but you might start there.

File: pkg/allocation/converters/converter_test.go (hunk 4 above)
Comment (Kalaiselvi84) [reply pkg/allocation/converters/converter_test.go:218]: I apologize for choosing the issue myself. I have restored the code and I am not sure what changes are required to correct the output. Could you please help me?

File: site/content/en/docs/Reference/gameserverallocation.md [hunk 7]
```diff
@@ -78,13 +78,7 @@ spec:
     matchLabels:
       game: my-game
     matchExpressions:
-      - {key: tier, operator: In, values: [cache]}
-    # [Stage:Beta]
-    # [FeatureFlag:StateAllocationFilter]
-    # Specifies which State is the filter to be used when attempting to retrieve a GameServer
-    # via Allocation. Defaults to "Ready". The only other option is "Allocated", which can be used in conjunction with
-    # label/annotation/player selectors to retrieve an already Allocated GameServer.    
-    gameServerState: Ready
```
Comment (gongmax) [inline site/content/en/docs/Reference/gameserverallocation.md:87]: This should not be removed.

Comment (gongmax) [reply site/content/en/docs/Reference/gameserverallocation.md:87]: Should also remove line 39 and 40

**Comment by gongmax** [comment]
Also need to remove 
- https://github.com/googleforgames/agones/blame/main/examples/gameserverallocation-deprecated.yaml#L40-L41
- https://github.com/googleforgames/agones/blob/main/examples/gameserverallocation.yaml#L49-L50

And please resolve the merge conflicts

File: site/content/en/docs/Reference/gameserverallocation.md (hunk 7 above)
Comment (Kalaiselvi84) [reply site/content/en/docs/Reference/gameserverallocation.md]: Restored the code and removed the lines 39 and 40.

**Comment by Kalaiselvi84** [comment]
> Also need to remove
> 
> * https://github.com/googleforgames/agones/blame/main/examples/gameserverallocation-deprecated.yaml#L40-L41
> * https://github.com/googleforgames/agones/blob/main/examples/gameserverallocation.yaml#L49-L50
> 
> And please resolve the merge conflicts

Resolved the conflicts👍🏻

**Pull request closed without merging**